plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/procedure/RegisterTableProcedure.java

```diff
@@ -0,0 +1,166 @@
+import static io.trino.plugin.deltalake.DeltaLakeMetadata.setDeltaStorageFormat;
+import static io.trino.plugin.deltalake.transactionlog.TransactionLogUtil.getTransactionLogDir;
+import static io.trino.plugin.hive.metastore.MetastoreUtil.buildInitialPrivilegeSet;
+import static io.trino.spi.StandardErrorCode.GENERIC_USER_ERROR;
+import static io.trino.spi.StandardErrorCode.PERMISSION_DENIED;
+import static io.trino.spi.StandardErrorCode.SCHEMA_NOT_FOUND;
+import static io.trino.spi.type.VarcharType.VARCHAR;
+import static java.lang.String.format;
+import static java.lang.invoke.MethodHandles.lookup;
+import static java.util.Objects.requireNonNull;
+import static org.apache.hadoop.hive.metastore.TableType.EXTERNAL_TABLE;
+
+public class RegisterTableProcedure
+        implements Provider<Procedure>
+{
+    private static final MethodHandle REGISTER_TABLE;
+
+    private static final String PROCEDURE_NAME = "register_table";
+    private static final String SYSTEM_SCHEMA = "system";
+
+    private static final String SCHEMA_NAME = "SCHEMA_NAME";
+    private static final String TABLE_NAME = "TABLE_NAME";
+    private static final String TABLE_LOCATION = "TABLE_LOCATION";
+
+    static {
+        try {
+            REGISTER_TABLE = lookup().unreflect(RegisterTableProcedure.class.getMethod("registerTable", ConnectorSession.class, String.class, String.class, String.class));
+        }
+        catch (ReflectiveOperationException e) {
+            throw new AssertionError(e);
+        }
+    }
+
+    private final DeltaLakeMetadataFactory metadataFactory;
+    private final TrinoFileSystemFactory fileSystemFactory;
+    private final boolean registerTableProcedureEnabled;
+
+    @Inject
+    public RegisterTableProcedure(DeltaLakeMetadataFactory metadataFactory, TrinoFileSystemFactory fileSystemFactory, DeltaLakeConfig deltaLakeConfig)
+    {
+        this.metadataFactory = requireNonNull(metadataFactory, "metadataFactory is null");
+        this.fileSystemFactory = requireNonNull(fileSystemFactory, "fileSystemFactory is null");
+        this.registerTableProcedureEnabled = requireNonNull(deltaLakeConfig, "deltaLakeConfig is null").isRegisterTableProcedureEnabled();
+    }
+
+    @Override
+    public Procedure get()
+    {
+        return new Procedure(
+                SYSTEM_SCHEMA,
+                PROCEDURE_NAME,
+                List.of(
+                        new Procedure.Argument(SCHEMA_NAME, VARCHAR),
+                        new Procedure.Argument(TABLE_NAME, VARCHAR),
+                        new Procedure.Argument(TABLE_LOCATION, VARCHAR)),
+                REGISTER_TABLE.bindTo(this));
+    }
+
+    public void registerTable(
+            ConnectorSession clientSession,
+            String schemaName,
+            String tableName,
+            String tableLocation)
+    {
+        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(getClass().getClassLoader())) {
```

Comment:
Aren't we supposed to apply the same pattern as for Iceberg ? 

https://github.com/trinodb/trino/blob/ccc06a5a8d2d2ab6b3df4045fe0cc08494744240/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/procedure/RegisterTableProcedure.java#L115

The class loader at the time when `registerTable` method is being called may be the app class loader which is not what we want.

Reply:
Yes, It should be the same pattern. Addressed comment.

Reply:
> It should be the same pattern.

agreed



> Aren't we supposed to apply the same pattern as for Iceberg ?

agreed



> The class loader at the time when `registerTable` method is being called may be the app class loader which is not what we want.

@findinpath 

no. `getClass().getClassLoader()` was fine here.
it doesn't matter whether you do `this.classLoader = getClass().getClassLoader();` in ctor
or only `getClass().getClassLoader()` later -- the answer is the same